java/arcs/core/util/Lens.kt

```diff
@@ -0,0 +1,187 @@
+ * val parent = Parent("Mary", listOf(Person("Jack", 4), Person("Jill", 3)))
+ * val childrenLens = lens(Parent::children) { t, f = t.copy(children = f) }
+ * val ageLens = lens(Person::age) { t,f -> t.copy(age = f) }
+ *
+ * val allChildren = childrenLens.traverse()
+ * val newParent = allChildren.mod(parent) { it.copy(name = "Baby ${it.name}") }
+ * // newParent = Parent("Mary", listOf(Person("Baby Jack", 4), Person("Baby Jill", 3)))
+ * ```
+ *
+ * ## Traversal Composition
+ * A [Traversal] can be combined with a [Lens], or a another [Traversal] to reach deep into
+ * nested data structures.
+ *
+ * ```kotlin
+ * val allChildrenAges = allChildren + ageLens
+ * val newParent = allChildrenAges.mod(parent) { it + 1 }
+ * // newParent = Parent("Mary", listOf(Person("Jack", 5), Person("Jill", 4)))
+ * ```
+ *
+ * ## Reference
+ * Based on the ideas in
+ * [http://davids-code.blogspot.com/2014/02/immutable-domain-and-lenses-in-java-8.html]
+ *
+ * @property getter a lambda which given [Target] returns [Focus]
+ * @property setter a lambda which given [Target] and [Focus] returns new [Target] set with [Focus]
+ */
+open class Lens<Target, Focus>(
+    private val getter: (Target) -> Focus,
+    private val setter: (Target, Focus) -> Target
+) {
+    /** Invokes a getter on the [Target] and returns a [Focus] value. */
+    fun get(target: Target) : Focus = getter(target)
+
+    /** Invokes a setter on the [Target] with the given [Focus] value and returns a new [Target]. */
+    fun set(target: Target, value: Focus) : Target = setter(target, value)
+
+    /**
+     * Applies a function to the value of invoking [get] on the [Target], and then invokes
+     * [set] on the [Target] with the value, returning a new [Target].
+     */
+    fun mod(target: Target, f: (Focus) -> Focus): Target {
+        return set(target, f(get(target)))
+    }
+
+    /**
+     * Allows two or more [Lens] to be composed and chained so that a deeply nested immutable
+     * data structure can be elegantly updated. Returns a new [Lens].
+     */
+    fun <Parent> comp(parentLens: Lens<Parent, Target>) : Lens<Parent, Focus> =
+        object : Lens<Parent, Focus>(
+            { parent : Parent -> get(parentLens.get(parent)) },
+            { parent : Parent, focus: Focus ->
+                parentLens.mod(parent) { target: Target ->
+                    set(target, focus)
+                }
+            }
+        ) {}
+}
+
+/**
+ * A [Traversal] is sort of a [Lens] with more than one focii. A [Lens] could be considered a
+ * [Traversal] with a single focii.
+ */
+interface Traversal<Target, Focus> {
+    /**
+     * Apply [f] to every focii in this traversal and return an updated target.
+     */
+    fun mod(target: Target, f: (Focus) -> Focus): Target
+}
+
+/**
+ * A [Traversal] over the [List] returned by a [Lens] whose focus type is a `List<Focus>`.
+ */
+class ListLensTraversal<Target, Focus>(
+    private val listLens: Lens<Target, List<Focus>>
```

Comment:
Could this just as well be `Lens<Target, Iterable<Focus>>`?

Reply:
Unfortunately no, I tried this in the beginning, but mapValues() doesn't work on iterables, which means the function would have to return Entry<K,V>

Reply:
Couldn't the `Map.traverse` helper function just call `IterableLensTraversal(map.values)` in that case?

Reply:
Its the Lens that you call traverse on (because the map is usually in the field of another object), so you'd need the Lens traverse to recognize when the focus is a map, e.g.

```
fun <Subject, Focus> Lens<Subject, Map<String, Focus>>.traverse = ...
```

But I think when I tried this, Kotlin's type inference didn't like it due to type-erasure problem.